[WIP] sets: Implements the indicator functions for set theory.

jgulian opened the pull request:
#### References to other Issues or PRs
Fixes #19481

#### Brief description of what is fixed or changed
Implements the indicator function (or characteristic function) which maps a number x
to the range {0, 1} for a set A; the function returns 1 if x is in A, and 0 if x is
not in A. This implementation works for defined numbers, and sets as well as symbolic
numbers and sets. The function, indicator, also includes the inverse of the indicator
function, the Mobius function (which has already been added for number theory). This
also adds relevant tests for the indicator function.

#### Other comments
Examples:
```python
In [1]: from sympy.sets import Union, Interval, FiniteSet, indicator
In [2]: set = FiniteSet(1, 2, 3)
In [3]: indicator(0, set)
Out [3]: 0
In [4]: indicator(1, set)
Out [4]: 1
```

References:
https://en.wikipedia.org/wiki/Indicator_function
https://mathworld.wolfram.com/CharacteristicFunction.html

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* sets
  * adds indicator function for sets
<!-- END RELEASE NOTES -->

jgulian committed:
sets: This implements the indicator functions for set theory.

Implements the indicator function (or characteristic function) which maps a number x
to the range {0, 1} for a set A; the function returns 1 if x is in A, and 0 if x is
not in A. This implementation works for defined numbers, and sets as well as symbolic
numbers and sets. The function, indicator, also includes the inverse of the indicator
function, the Mobius function (which has already been added for number theory). This
also adds relevant tests for the indicator function.

Examples:

In [1]: from sympy.sets import Union, Interval, FiniteSet, indicator
In [2]: set = FiniteSet(1, 2, 3)
In [3]: indicator(0, set)
Out [3]: 0
In [4]: indicator(1, set)
Out [4]: 1

References:
https://en.wikipedia.org/wiki/Indicator_function
https://mathworld.wolfram.com/CharacteristicFunction.html

oscarbenjamin commented on sympy/sets/indicator.py:
```diff
@@ -0,0 +1,58 @@
+    def eval(cls, x, a: Set):
+        if not isinstance(x, Symbol):
+            if type(a) != Set:
+                return 1 if a.contains(x) else 0
```
We don't need these if conditions. Also `contains` has three different possible returns true, false, and indeterminate:
```julia
In [3]: Interval(0, 1).contains(0)                                                                                                             
Out[3]: True

In [4]: Interval(0, 1).contains(2)                                                                                                             
Out[4]: False

In [5]: Interval(0, 1).contains(Symbol('x'))                                                                                                   
Out[5]: 0 ≤ x ∧ x ≤ 1
```
This function would return 0 in the 3rd case. It should be something like:
```python
is_in_set = a._contains(x)
if is_in_set is True:
    return 1
elif is_in_set is False:
    return 0
# None is implicitly returned otherwise
```

oscarbenjamin commented on sympy/sets/indicator.py:
```diff
@@ -0,0 +1,58 @@
+        """
+        Returns the Mobius function, the inverse of the indicator function. [1]
+        """
+        return mobius
```
This function is not invertible in any normal sense so it shouldn't have an `inverse` method.

oscarbenjamin commented on sympy/sets/indicator.py:
```diff
@@ -0,0 +1,58 @@
+    .. [2] https://mathworld.wolfram.com/CharacteristicFunction.html
+
+    """
+
```
There could be class attributes here:
```
is_integer = True
is_nonnegative = True
is_prime = False
```

oscarbenjamin commented on sympy/sets/tests/test_indicator.py:
```diff
@@ -0,0 +1,25 @@
+
+
+def test_indicator_symbolic_function():
+    A = Set('A')
```
We shouldn't use `Set('A')` like this. It doesn't really represent a symbolic set. The `Set` class is essentially abstract. There should be a `SetSymbol` class but there isn't.

oscarbenjamin commented:
I wonder if it makes more sense to have something like
```
f1 = indicator(Interval(2, 3))
f1(2)  #  evaluates to 1
```

asmeurer commented on sympy/sets/indicator.py:
```diff
@@ -0,0 +1,58 @@
+    @classmethod
+    def eval(cls, x, a: Set):
+        if not isinstance(x, Symbol):
+            if type(a) != Set:
```
Use isinstance instead of type comparisons. This will fail for subclasses of Set.

asmeurer replied:
Maybe it makes sense to have a type check like

```py
if not isinstance(a, Set):
    raise ValueError("The second argument of indicator must be a set")
```

asmeurer commented on sympy/sets/tests/test_indicator.py:
```diff
@@ -0,0 +1,25 @@
+    assert indicator(1, C) == 1
+
+
+def test_indicator_symbolic_function():
```
You can use the `unchanged` function in this test to test that it doesn't evaluate.

asmeurer commented on sympy/sets/indicator.py:
```diff
@@ -0,0 +1,58 @@
+
+    It is defined as follows:
+        1) `1` if x exists in A
+        2) `0` if x does not exist in A
```
We can write the definition as a piecewise using LaTeX

asmeurer commented:
I would just put this in sets.py instead of a new file.

asmeurer commented on sympy/sets/indicator.py:
```diff
@@ -0,0 +1,58 @@
+    ========
+
+    >>> from sympy.sets import Union, Interval, FiniteSet, indicator
+    >>> set = Union(Interval(0, 1), Interval(2, 3))
```
Don't use `set` as a variable name. This overwrites the built-in `set`, so it's a bad practice.

asmeurer commented on sympy/sets/indicator.py:
```diff
@@ -0,0 +1,58 @@
+    """
+
+    @classmethod
+    def eval(cls, x, a: Set):
```
I would call the second argument capital `A`.

asmeurer replied:
`x` can be type `Expr`

asmeurer replied:
Actually x is better as type Basic, since sets can contain things other than Expr.

asmeurer commented on sympy/sets/indicator.py:
```diff
@@ -0,0 +1,58 @@
+
+    @classmethod
+    def eval(cls, x, a: Set):
+        if not isinstance(x, Symbol):
```
x can be any expression. There's no need to do something different for just a single symbol.

asmeurer commented:
I think it would be good to have LaTeX and pretty-printing for this. For Unicode pretty printing, I think the best choice is 𝟙 (MATHEMATICAL DOUBLE-STRUCK DIGIT ONE, see https://en.wikipedia.org/wiki/Mathematical_operators_and_symbols_in_Unicode).

asmeurer commented:
> I wonder if it makes more sense to have something like

That would be a lot more complicated to implement. IMO currying isn't really necessary in SymPy since you can always use a dummy symbol to represent an undefined function. Although I do wonder if `indicator(A, x)` is a more natural argument order, since the notation is written as `1_A(x)`. Either way, we could raise a helpful error message if the order is backwards, since the types won't be the same.

asmeurer commented:
I can't find any examples of indicators functions from other CASes or languages to compare to to see what argument order they use. I guess it's because in most programming languages booleans are integers, so "x in A" or the equivalent is sufficient as an indicator function.

asmeurer commented:
If we implement the equivalent of this for booleans (I don't know if that should still be called "indicator"), then this can be implemented using that and `Contains`. Obviously it's not a complicated implementation either way, but I will note that the boolean version of this is much more needed than the set version.

sylee957 commented:
Why does it need to be implemented for booleans?
I see there is already `Contains`, so `Piecewise((1, Contains(x, S)), (0, True))` should be sufficient.

asmeurer commented:
Because it's a common thing to need to convert between them, and writing `Piecewise((1, boolean), (0, True))` is tedious.

jgulian committed:
sets: This improves upon the design of the indicator function

Moved indicator function from seperate file into sets.py of the sets package
(along with the tests into test_sets.py). Redefines the indicator function in
terms of latex. Changes the order of the function to (A: Set, x: Basic) from
(x: Basic, A: Set). Adds latex printing for indicators as well as tests.

jgulian commented:
I made most of the changes. 

I did not know where to put the boolean indicator (or what to call it). I thought it seemed to make sense in logic/boolalg.py as a globally defined Piecewise function to easily import into logic/__init__.py; however Piecewise is dependent on boolalg, so importing Piecewise in boolalg gives a circular import error. Also, I was unsure how to implement pretty printing for indicators, so I couldn't add that. Some guidance in these two areas would be appreciated.

Thank you

oscarbenjamin commented:
The quality test failed:
```
AssertionError: File ends with more than 1 newline: /home/travis/build/sympy/sympy/sympy/sets/sets.py, line 2572
```

oscarbenjamin commented:
I would add this in the sets module somewhere.

For printing if the class is called `indicator` then you need to add methods like `_print_indicator` to the printer classes. You can see e.g. where the methods are for FiniteSet with:
```
$ git grep _print_FiniteSet
sympy/printing/latex.py:    def _print_FiniteSet(self, s):
sympy/printing/mathml.py:    def _print_FiniteSet(self, e):
sympy/printing/mathml.py:    def _print_FiniteSet(self, s):
sympy/printing/pretty/pretty.py:    def _print_FiniteSet(self, s):
sympy/printing/tests/test_mathml.py:def test_print_FiniteSet():
```

jgulian committed:
sets: implements boolean_indicator and printing for both indicators

Implements an indicator function for booleans which maps true values to 1,
and false values to 0. Both latex, and pretty printing are supported for
the boolean indicator as well as the regular indicator. Relevant tests are
also added.

Example:

In[1]: from sympy.sets import boolean_indicator
In[2]: from sympy import true, false
In[3]: boolean_indicator(True)
Out[3]: 1
In[4]: boolean_indicator(False)
Out[4]: 0
In[5]: boolean_indicator(true)
Out[5]: 1
In[6]: boolean_indicator(false)
Out[6]: 0

jgulian commented:
I see it failed two tests, and I am working on it.

jgulian committed:
core: add indicator and boolean_indicator from sets to test_args

oscarbenjamin commented:
If we have the `boolean_indicator` function then do we also need the `indicator_function`. With `boolean_indicator` we can turn any boolean predicate into a `0/1`-valued expression. Applying that to `Contains` gives us the same effect as `indicator_function`. Is it necessary to have more than just `boolean_indicator` and is that a good name?

jgulian commented:
I can remove the set one, and rename the boolean indicator to indicator, if you want.

sylee957 commented:
I have doubts whether it’s going to be easy to parse such composed indicator function when building symbolic integrals or logic solvers, if you mean to parse stuff like ‘bool_to_int(bool_indicator(x, S))’
So I don’t think that it should be more centered for logic than set function. At least for logics, we can still settle down with Piecewise because it has lots of methods to simplify itself.

oscarbenjamin commented:
> So do you think that ‘boolean_indicator(Contains(x, S))’ is going to be more easier for users than Piecewise?

I'm not sure how many users find `Piecewise` easy to use. I have to say it's not clear to me what either of the `indicator` functions here would be used for.

jgulian commented:
It was for issue #19481.

sylee957 commented on sympy/sets/sets.py:
```diff
@@ -2516,3 +2517,106 @@ def set_pow(x, y):
+    def eval(cls, A: Set, x: Basic):
+        x_in_a = A.contains(x)
+        if x_in_a is true:
+            return 1
```
These returns should be sympified to give S.One and S.Zero instead of python 1 and 0 respectively.

jgulian committed:
sets: Changes the returns of the indicator functions

The indicator functions initially used 0, and 1; this commit changes them
to use S.Zero, and S.One.
